### PR TITLE
Test/test full text search

### PR DESCRIPTION
**Pull request by rajabi-m**

### CODE REVIEW

File: 04-unit-test/src/App/Services/FileLoader.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Mohaymen.FullTextSearch.App.Services;
+
+public class FileLoader
+{
+    private IFileReader _fileReader;
+
+    public FileLoader(IFileReader fileReader)
+    {
+        _fileReader = fileReader;
```
Comment (AliSK81) [inline 04-unit-test/src/App/Services/FileLoader.cs:16]: check for arugment null exception

File: 04-unit-test/test/App/Utilities/InputParserTest.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        var queries = _inputParser.ParseToSearchQuery(input);
+
+        // Assert
+        var mandatories = queries.Find(query => query.SearchStrategy is MandatorySearchStrategy)?.Keywords;
+        var optionals = queries.Find(query => query.SearchStrategy is OptionalSearchStrategy)?.Keywords;
+        var excludeds = queries.Find(query => query.SearchStrategy is ExcludedSearchStrategy)?.Keywords;
+
+        Assert.Equal(expectedMandatories, mandatories);
+        Assert.Empty(optionals);
+        Assert.Empty(excludeds);
```
Comment (AliSK81) [inline 04-unit-test/test/App/Utilities/InputParserTest.cs:142]: بهتر بود 3 تا تست بشه

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/OptionalSearchStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        // Arrange
+        var documents = new HashSet<string>(_invertedIndex.AllDocuments);
+        var keywords = new List<Keyword> { new Keyword("academy"), new Keyword("summer") };
+
+        // Act
+        _searchStrategy.FilterDocuments(documents, keywords, _invertedIndex);
+
+        // Assert
+        var expected = new HashSet<string> { "doc1.txt", "doc3.txt", "doc4.txt" };
+        Assert.True(expected.SetEquals(documents), $"Expected: {string.Join(", ", expected)}, Actual: {string.Join(", ", documents)}");
```
Comment (AliSK81) [inline 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/OptionalSearchStrategy.cs]: خوانایی توی تست ها هم مهمه

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/MandatorySearchStrategyTest.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                .Returns(entry.Value);
+        }
+
+        invertedIndex.AllDocuments.Returns(["doc1.txt", "doc2.txt", "doc3.txt", "doc4.txt"]);
+
+        return invertedIndex;
+    }
+    
+    [Fact]
+    public void FilterDocuments_ShouldIncludeOnlyDocumentsContainingAllKeywords()
```
Comment (AliSK81) [inline 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/MandatorySearchStrategyTest.cs:43]: اسم تست ها میتونه دقیق تر بشه
مثل یه بخش سومی که یک state رو نشون بده

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/ExcludedSearchStrategyTest.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        // Arrange
+        var documents = new HashSet<string>(_invertedIndex.AllDocuments);
+        var keywords = new List<Keyword> { new Keyword("star") };
+
+        // Act
+        _searchStrategy.FilterDocuments(documents, keywords, _invertedIndex);
+
+        // Assert
+        var expected = new HashSet<string> { "doc3.txt", "doc4.txt" };
```
Comment (AliSK81) [inline 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/ExcludedSearchStrategyTest.cs]: this is arrange

**Review by AliSK81 — APPROVED** [review]
خسته نباشید

File: 04-unit-test/test/App/Utilities/InputParserTest.cs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        [
+            "",
+            new List<Keyword> { },
+            new List<Keyword> { },
+            new List<Keyword> { },
+        ];
+
+        yield return
+        [
+            "hamed dooset -darim",
```
Comment (HamedSY) [inline 04-unit-test/test/App/Utilities/InputParserTest.cs:23]: :))))))))

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/InvertedIndexSearcherTest.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            new SearchQuery(optionalSearchStrategy, optionalKeywords),
+            new SearchQuery(excludedSearchStrategy,  excludedKeywords)
+        };
+
+        // Act
+        var results = _searcher.Search(queries);
+
+        // Assert
+        var expected = new HashSet<string> { "doc3.txt" };
+        Assert.True(expected.SetEquals(results), $"Expected: {string.Join(", ", expected)}, Actual: {string.Join(", ", results)}");
```
Comment (HamedSY) [inline 04-unit-test/test/DocumentManagement/Services/InvertedIndex/InvertedIndexSearcherTest.cs:66]: این مدل نوشتن لازمه؟ مسیج خود ارورا مگه خوانا و خوب نیست؟

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            .Returns([new Keyword("summer")]);
+        
+        // Act
+        IInvertedIndex invertedIndex = _filesInvertedIndexBuilder.IndexFilesWords(fileCollection).Build();
+        
+        // Assert
+        Assert.Equal(["doc1.txt", "doc2.txt"], invertedIndex.GetDocumentsByKeyword(new Keyword("star")));
+        Assert.Equal(["doc1.txt", "doc3.txt"], invertedIndex.GetDocumentsByKeyword(new Keyword("academy")));
+        Assert.Equal(["doc2.txt", "doc3.txt"], invertedIndex.GetDocumentsByKeyword(new Keyword("coder")));
+        Assert.Equal(["doc4.txt"], invertedIndex.GetDocumentsByKeyword(new Keyword("summer")));
```
Comment (HamedSY) [inline 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs:45]: این فیلم رو درباره fluent assertion ببینید.
https://www.youtube.com/watch?v=b2zxl5zNjlA

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    private readonly FilesInvertedIndexBuilder _filesInvertedIndexBuilder;
+    
+    public FilesInvertedIndexBuilderTest()
+    {
+        _tokenizer = Substitute.For<ITokenizer>();
+        _filesInvertedIndexBuilder = new FilesInvertedIndexBuilder(_tokenizer);
+    }
+
+    [Fact]
+    public void IndexFilesWords_ValidFileCollection_ShouldIndexAllWords()
```
Comment (HamedSY) [inline 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs]: تو همون فیلمی که از fluent assertion گذاشتم به اسمای توابع تستش دقت کنید. یه قسمت when هم داره که شرایط اون تست رو میگه. خوبه اونم تو اسمای تستاتون اضافه کنید

**Review by HamedSY — APPROVED** [review]

File: 04-unit-test/test/App/Utilities/InputParserTest.cs (hunk 2 above)
Comment (rajabi-m) [reply 04-unit-test/test/App/Utilities/InputParserTest.cs:142]: توی این تست فقط استراتژی mandatory داره تست میشه و فقط چک کردم که اونا خالی باشن.

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/MandatorySearchStrategyTest.cs (hunk 4 above)
Comment (rajabi-m) [reply 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/MandatorySearchStrategyTest.cs:43]: چون برای این تابع حالت edge case خاصی پیدا نکردیم فقط یک تست براش نوشتیم و توی این تست هم state خاصی برسی نمیشه و ورودی ها کلی هستند.

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/InvertedIndexSearcherTest.cs (hunk 7 above)
Comment (rajabi-m) [reply 04-unit-test/test/DocumentManagement/Services/InvertedIndex/InvertedIndexSearcherTest.cs:66]: توی این حالت مجبور بودیم از تابع SetEqual هش‌ست استفاده کنیم و برای همین assert.Equal نزدیم و محبور شدیم خودمون هم یه ارور خوب براش بنویسیم.

File: 04-unit-test/test/App/Utilities/InputParserTest.cs (hunk 2 above)
Comment (rajabi-m) [reply 04-unit-test/test/App/Utilities/InputParserTest.cs:142]: توی این تست فقط استراتژی mandatory داره تست میشه و فقط چک کردم که اونا خالی باشن.

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs (hunk 8 above)
Comment (rajabi-m) [reply 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs:45]: دیدیم ممنون

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs (hunk 9 above)
Comment (rajabi-m) [reply 04-unit-test/test/DocumentManagement/Services/InvertedIndex/FilesInvertedIndexBuilderTest.cs]: اصلاح شد

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/ExcludedSearchStrategyTest.cs (hunk 5 above)
Comment (rajabi-m) [reply 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/ExcludedSearchStrategyTest.cs]: اصلاح شد

File: 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/OptionalSearchStrategy.cs (hunk 3 above)
Comment (rajabi-m) [reply 04-unit-test/test/DocumentManagement/Services/InvertedIndex/SearchStrategies/OptionalSearchStrategy.cs]: اصلاح شد

File: 04-unit-test/src/App/Services/FileLoader.cs (hunk 1 above)
Comment (rajabi-m) [reply 04-unit-test/src/App/Services/FileLoader.cs:16]: اصلاح شد
